redesign: Improve multi-threaded domain performance

snadampal opened the pull request:
1. Grouped ofi ep,av, and cq together and made them part of plugin EndPoint class
2. Maintained a per thread plugin EndPoint object
3. Added support for multiple domains per device based on application request
4. Updated the nccl net api interface to pass domain key to request for a new domain
5. Enabled 'FI_THREAD_COMPLETION' threading mode because now the completion
structures are created per thread.
6. Made Plugin endpoint ops thread safe to support multiple comms per thread scenario
7. Added endpoint state tracking to handle abrupt comm abort scenarios
8. Fixed `nccl_connect` test application to avoid race condition
   between the comm creation and closing between the sender and receiver ranks.
9. Implemented `ep_per_comm` feature to support rComms have a dedicated endpoint
but shared CQ.
10. Implemented an endpoint aware MR cache policy to support providers with FI_MR_ENDPOINT caps

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

bwbarrett commented on src/nccl_ofi_api.cpp:
```diff
@@ -251,6 +251,13 @@ ncclResult_t nccl_net_ofi_listen_v2(int dev, void* handle, void** listenComm)
+       return nccl_net_ofi_listen_v11(dev_id, handle, lComm, 0, 0);
+}
+
+ncclResult_t nccl_net_ofi_listen_v11(int dev_id, void *handle, void **lComm,
```
I feel somewhat compelled to squawk that this should be its own commit.

bwbarrett commented on src/nccl_ofi_rdma.cpp:
```diff
@@ -1696,8 +1695,17 @@ int nccl_net_ofi_rdma_ep_t::ofi_process_cq()
+	for (uint16_t rail_id = 0; rail_id != this->num_rails; ++rail_id) {
+		nccl_net_ofi_ep_rail_t *rail = this->rdma_endpoint_get_rail(rail_id);
+
+		ret = ofi_process_cq_rail(device, rail);
```
I thought we were still going to share the CQ between control and data rails.  Don't love having to poll each CQ twice, because cq polling performance has been a constant thorn in our side.

snadampal replied:
thanks, Brian. I will revisit the cq process function for concurrent accesses and the ep/cq objects.

snadampal replied:
Hi @bwbarrett , I have incorporated your feedback and updated the PR.

rauteric commented on src/nccl_ofi_net.cpp:
```diff
@@ -831,23 +812,45 @@ int nccl_net_ofi_device_t::release_all_domain_and_ep()
-	if (ep == nullptr) {
+	auto ep_iter = this->ep_table.find(lookup_key);
+
+	if (ep_iter != this->ep_table.end()) {
```
Nit: indenting in this section is off.

snadampal commented:
I have updated the PR for an "endpoint aware MR cache policy" to support `FI_MR_ENDPOINT` providers  and `endpoint_per_communicator` features.

rauteric commented on include/nccl_ofi_rdma.h:
```diff
@@ -1043,17 +1050,43 @@ class nccl_net_ofi_rdma_ep_t : public nccl_net_ofi_ep_t {
+	 */
+	inline nccl_net_ofi_rdma_cq_rail_t *rdma_endpoint_get_cq_rail(uint16_t rail_id)
+	{
+		if (parent_endpoint)
```
On reading this code, I don't really like the `parent_endpoint` pattern and conditionals like this. It feels like this is a consequence of trying to do two things in the endpoint class (endpoint and cq management), instead of having separate classes.

snadampal replied:
yes, that's true, because of CQ is now part of Plugin Endpoint class.

rauteric commented on src/nccl_ofi_rdma.cpp:
```diff
@@ -6605,6 +6669,10 @@ nccl_net_ofi_rdma_ep_t::nccl_net_ofi_rdma_ep_t(nccl_net_ofi_rdma_domain_t *domai
 		throw std::runtime_error("rdma endpoint constructor: initializing rx_buffers failed");
 	}
+
+	/* Connection manager for this endpoint */
```
Note: we don't need a connection manager for the endpoints with a parent_ep.

rauteric commented on src/nccl_ofi_rdma.cpp:
```diff
@@ -6471,6 +6501,24 @@ int nccl_net_ofi_rdma_ep_t::release_ep(bool skip_lock, bool force_cleanup)
 			}
 			ret = this->cleanup_resources();
+
+			if (this->parent_endpoint) {
```
Won't this always be true in this path, because this is under `if (this->is_endpoint_per_communicator_ep)`?

rauteric commented on src/nccl_ofi_rdma.cpp:
```diff
@@ -6471,6 +6501,24 @@ int nccl_net_ofi_rdma_ep_t::release_ep(bool skip_lock, bool force_cleanup)
+					nccl_net_ofi_mutex_unlock(&domain_ptr->domain_lock);
+				}
+
+				this->parent_endpoint->release_ep(true, false);
```
Why set `skip_lock=true` here? I don't see that we are holding any lock at this point in the code?

snadampal replied:
thanks, Eric. I will fix it.

rauteric commented on include/nccl_ofi_mr.h:
```diff
@@ -112,13 +118,15 @@ static inline void nccl_ofi_mr_ckey_round(size_t *len, void **base_addr, const c
 #if HAVE_DECL_FI_MR_DMABUF
-static inline nccl_ofi_mr_ckey_t nccl_ofi_mr_ckey_mk_dmabuf(int fd, uint64_t offset, size_t len, void *base_addr)
+static inline nccl_ofi_mr_ckey_t nccl_ofi_mr_ckey_mk_dmabuf(int fd, uint64_t offset, size_t len,
+							    void *base_addr, nccl_net_ofi_ep_t *ep_ptr)
```
This may not be sufficient for when we implement `endpoint_mr` for RDMA protocol. There, for a given endpoint there will be two distinct memory registrations: for control rails and for data rails. We would need to distinguish that here. (In practice, we shouldn't need to register the same memory region with both control and data rails, but some of our round-up behavior makes me nervous.)

rauteric reviewed (commented):
Left a few comments. Some are stylistic; I don't feel strongly we need to address those before merging.

snadampal commented:
Hi @bwbarrett , @rauteric , I have addressed the last few comments and rebased the PR to the mainline latest. Could you please do a final review and approve. Thank you!

rauteric commented on src/nccl_ofi_rdma.cpp:
```diff
@@ -2374,10 +2373,13 @@ static int test(nccl_net_ofi_req_t *base_req, int *done, int *size)
-	nccl_net_ofi_rdma_domain_t *domain = ep->rdma_endpoint_get_domain();
-	pthread_wrapper domain_lock(&domain->domain_lock);
+	pthread_mutex_t *ep_local_lock = &ep->ep_lock;
+	if (ep->is_endpoint_per_communicator_ep && ep->parent_endpoint)
```
This check is repeated many times. Does it make sense to just make the base `ep->ep_lock` a pointer, and set it to point to the parent for children EPs? (Or, make this a macro?)

snadampal replied:
thanks @rauteric . I thought the same but didn't do it because it's not the case for every lock, I see few places where we only use the child EP lock, for example abort(). 
But I will add the marco option to the todo list and do it along with any other fixes we may need to make when we take this plugin through the new test suite under development.

rauteric replied:
Actually, the invalidation is really supposed to be at the CQ level; the thought behind invalidation is that once we close an endpoint, it will generate completion events for the receives posted to the endpoint, and so it is more likely for the CQ to overflow (and we don't have overflow detection today). That's why today invalidation is done at domain level.

In this code, we should really be invalidating the parent endpoint for children EPs, since that's where the CQ is.
